src/platform_impl/linux/wayland/window/shim.rs

```diff
@@ -304,6 +311,27 @@ impl WindowHandle {
         }
     }
 
+    pub fn set_ime_allowed(&self, allowed: bool, event_sink: &mut EventSink) {
+        if self.ime_allowed.get() == allowed {
+            return;
+        }
+
+        self.ime_allowed.replace(allowed);
+        let window_id = wayland::make_wid(self.window.surface());
+
+        for text_input in self.text_inputs.iter() {
+            text_input.set_input_allowed(allowed);
+
+            let event = if allowed {
+                WindowEvent::IME(IME::Enabled)
+            } else {
+                WindowEvent::IME(IME::Disabled)
+            };
+
+            event_sink.push_window_event(event, window_id);
```

Comment:
Why is an event generated for every single IME? Shouldn't the event only be generated once?

Reply:
How so? If text input loses focus on the surface you send disable, since you don't have IME at all, meaning that IME is disabled, and when you get focus back you send enable?

Or are you talking in the context of multiple text-inputs?

Reply:
I'm talking in the context of multiple `text-inputs`. That's opaque to the client, right? But this still duplicates the event for all of the inputs. It doesn't seem useful to get the event `X` times? The `text_input` is never used for this, so it's not like you're passing the specific input?

Reply:
Ah, no, multiple text inputs are kind of broken right now, since I think protocol get clarified wrt that? At least I don't remember a lot of stuff from it, but the thing is that you can't have multiple text inputs over the same surface from what I can see for the same seat.

I'll try updating it according to the spec on Wayland side.

Reply:
I mean regardless of if it works properly or not, the code doesn't make any sense? Saying "this is broken anyway" seems like a weak excuse. I'd assume you have it as a `Vec` because it might work in the future? So why not write it properly? Just gotta move things down a couple lines.

Reply:
For whatever reason I've thought that you're talking about completely different place in handler, but it was in window/shim.rs. I've moved the lines, should be good.

I've also checked the `text-input` protocol, and what winit is doing make sense, since it has one text-input per seat, but the protocol talking about multiple text inputs per seat, so we're good.